src/main/java/org/mitre/synthea/world/concepts/Terminology.java

```diff
@@ -0,0 +1,575 @@
+import java.util.Random;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+
+import okhttp3.FormBody;
+import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+
+import org.hl7.fhir.dstu3.model.ValueSet;
+import org.mitre.synthea.helpers.Config;
+import org.mitre.synthea.world.concepts.HealthRecord.Code;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.InputSource;
+
+
+
+
+
+
+/**
+ * Finds a specified value set, extracts the codes and systems
+ * and selects one code/system at random.
+ */
+
+public class Terminology {
+
+  private static String VS_FOLDER = "valuesets";
+
+
+
+  private static final String VSAC_USER = Config.get("terminology.username");
+  private static final String VSAC_PASS = Config.get("terminology.password");
+
+  private static final String AUTH_URL = "http://vsac.nlm.nih.gov/vsac/ws/Ticket";
+
+  private static final String SERVICE_URL = "http://umlsks.nlm.nih.gov";
+
+  private static final String VSAC_VALUE_SET_RETRIEVAL_URL = "vsac.nlm.nih.gov";
+  private static final FhirContext ctx = FhirContext.forDstu3();
+  private static final Map<String, ValueSet> valueSets = loadValueSets();
+  private static final Map<String,String> codeLookup = loadLookupTable();
+
+  private static final Logger logger = LoggerFactory.getLogger(Terminology.class);
+  private static boolean connectionFailed = false;
+
+
+  public static Session sess;
+
+  static {
+    sess = new Session();
+  }
+
+
+
+  public static class Session {
+
+    private OkHttpClient client;
+
+    // Can be reused for multiple requests
+    private String authToken;
+
+
+    /**
+     * Sessions hold on to the client and auth token so that we don't have to
+     * keep making them, but you can make multiple instances in case you need
+     * more than one type of client.
+     */
+
+    public Session() {
+
+      if (VSAC_PASS == null | VSAC_USER == null) {
+        authToken = null;
+      } else if (!connectionFailed) {
+        client = getClient();
+        authToken = Terminology.getAuthToken(client);
+
+      }
+
+
+    }
+
+    /**
+     * Retrieves a value set from file, and if it can't find one, looks on VSAC.  Returns null if
+     * it can't find anything.
+     * @param url the uri, oid, or url of the value set
+     * @return the value set that was requested, either from file or VSAC
+     */
+    ValueSet getValueSet(String url) {
+      // Call to get valueSet from a Map
+      ValueSet vs = valueSets.get(url);
+      if (vs == null & authToken != null) {
+        String ticket = getServiceTicket(authToken,client);
+        Response response = Terminology.getValueSet(client,url,ticket);
```

Comment:
I'm a little confused on the relationship between the Session class and the Terminology class -- are there really two distinct concepts here? Is it possible to move the methods from one into the other and simplify things?

Reply:
The session was originally to make an object that had access to the auth token and client since giving the Terminology class access to the auth token might have caused problems if you wanted two different clients.  Also, some classes might not need or want the connection to VSAC, so the Session class used to hold only methods directly related to getting info from VSAC.

Looking at the code now, there's not much of a reason to keep them separated.  I'll make all the methods part of the Terminology class and give it some member variables for the token and client.